lib/java/src/org/apache/thrift/partial/PartialThriftDeserializer.java

```diff
@@ -0,0 +1,337 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.thrift.partial;
+
+import org.apache.thrift.partial.Validate;
+
+import org.apache.thrift.TBase;
+import org.apache.thrift.TEnum;
+import org.apache.thrift.TException;
+import org.apache.thrift.TFieldIdEnum;
+import org.apache.thrift.meta_data.EnumMetaData;
+import org.apache.thrift.meta_data.StructMetaData;
+import org.apache.thrift.protocol.TList;
+import org.apache.thrift.protocol.TMap;
+import org.apache.thrift.protocol.TSet;
+import org.apache.thrift.protocol.TType;
+
+/**
+ * Supports partial deserialization of serialized Thrift data.
+ *
+ * The end result of deserialization can be a Thrift object (? extends TBase)
+ * or it could be a different construct. It is achieved by using a
+ * ThriftFieldValueProcessor instance which processes each field value
+ * as it gets deserialized.
+ */
+public class PartialThriftDeserializer<T extends TBase> {
+
+  // Metadata that describes fields to deserialize.
+  private final ThriftMetadata.ThriftStruct metadata;
+
+  // Processor that handles deserialized field values.
+  private final ThriftFieldValueProcessor processor;
+
+  // Partial thrift protocol to use for deserialization.
+  private final PartialThriftProtocol protocol;
+
+  /**
+   * Constructs an instance of PartialThriftDeserializer.
+   *
+   * @param metadata the Metadata that describes fields to deserialize.
+   * @param processor the Processor that handles deserialized field values.
+   * @param protocol the Partial thrift protocol to use for deserialization.
+   */
+  public PartialThriftDeserializer(
+      ThriftMetadata.ThriftStruct metadata,
+      ThriftFieldValueProcessor processor,
+      PartialThriftProtocol protocol) {
+
+    Validate.checkNotNull(metadata, "metadata");
+    Validate.checkNotNull(processor, "processor");
+    Validate.checkNotNull(protocol, "protocol");
+
+    this.metadata = metadata;
+    this.processor = processor;
+    this.protocol = protocol;
+  }
+
+  /**
+   * Gets the Thrift metadata used by this instance.
+   */
+  public ThriftMetadata.ThriftStruct getMetadata() {
+    return this.metadata;
+  }
+
+  /**
+   * Deserializes the given serialized blob.
+   *
+   * @param bytes the serialized blob.
+   * @return deserialized instance.
+   * @throws TException if an error is encountered during deserialization.
+   */
+  public Object deserialize(byte[] bytes) throws TException {
```

Comment:
It returns Object on purpose. See the class header comment as well as the README.md file for more info.
This arrangement allows non-TBase values to be directly instantiated from serialized Thrift objects. One example of its usefulness is Spark engine reading SequenceFile of Thrift objects. In that case, this arrangement allows SparkThriftRowProcessor (not a part of this change) to directly deserialize a serialized Thrift blob into Spark's InternalRow representation. That enables significant performance boost.

Reply:
I'm still not convinced that this API need to be so dramatically different from the existing TDeserializer API (https://www.javadoc.io/doc/org.apache.thrift/libthrift/latest/org/apache/thrift/TDeserializer.html), can you try to match that as much as possible? The current TDeserializer API in Java already provided `partiallyDeserialize*` functions.

I'm really worried about how this API will map to other languages, even if it makes sense for the Java case.

Reply:
Hi @fishy, 
Happy to make changes if I get clarity on the concerns you have and on the type of changes you are looking for. 

Based on your comments I believe you prefer adding functionality to existing interfaces instead of adding new ones (please confirm since this is my guess). If so, how does the following work for you?

1. move common skip*() methods from PartialThriftProtocol.java to TProtocol.java.
2. move encoding specific skip*() methods from PartialThrift*Protocol.java to T*Protocol (for example, from PartialThriftBinaryProtocol.java to TBinaryProtocol)
3. get rid of PartialThriftDeserializer.java and add its methods to TDeserializer

Would something like the above address your concerns?